File: Introduction to Data Structures: Array and String/2. Introduction to 2D Array/Diagonal Traverse/DiagonalTraverse.py
class Solution:
    def findDiagonalOrder(self, mat: List[List[int]]) -> List[int]:

        # edge case: if matrix = 0
        if len(mat[0]) == 0:
            return mat

        # keep track of result
        result = []

        # set coordinates and variables
        x = 0
        y = 0
        p = len(mat) * len(mat[0])  # how many times we need to iterate

        while p > 0:

            # append current value at mat[x][y]
            result.append(mat[x][y])

            # if x + y coord is even
            if (x + y) % 2 == 0:

                # if you can go up and right
                if x - 1 >= 0 and y + 1 < len(mat[0]):

                    # go up
                    x -= 1

                    # go right
                    y += 1

                # else if you can't go up but can go right
                elif x - 1 < 0 and y + 1 < len(mat[0]):

                    # go right
                    y += 1

                # if you can't do both
                else:  # mat[x][y+1] != null and mat[x-1][y] != null:

                    # go down (edge)
                    x += 1

            # else if x + y is odd
            elif (x + y) % 2 != 0:

                # if you can go down and if you can go left
                if x + 1 < len(mat) and y - 1 >= 0:

                    # go down
                    x += 1

                    # go left
                    y -= 1

                # if you can go down but can't go left
                elif x + 1 < len(mat) and y - 1 < 0:

                    # go down
                    x += 1

                # if you cant do both
                else:

                    # go right
                    y += 1

            # dec p
            p -= 1

        return result

File: Introduction to Data Structures: Array and String/2. Introduction to 2D Array/Diagonal Traverse/DiagonalTraverse.java
class Solution {
    public int[] findDiagonalOrder(int[][] mat) {
       
        // input : matrix 
        // output: int [] arr
        
        // edge case when mat size is 0 
        if (mat.length == 0) {
            return new int [0];
        }
        
        // initialize variables: x and y coord, # times we need to iterate, result
        int x = 0;
        int y = 0;
        int l = mat.length * mat[0].length;
        int [] result = new int [l]; 
        
        // while number of times we need to iterate > 0 
        for (int i = 0; i < l; i++) {
        
            // append current matrix at [x][y] to the result array
            result[i] = mat[x][y];
        
            // case1: if x + y is even
            if ( (x + y) % 2 == 0 ) {
        
                // case: if you can go up and right
                if ( (x - 1) >= 0 && (y + 1) < mat[0].length ) {
                    x--; // go up
                    y++; // go right
                }
        
                // case: if you can't go up but can go right
                else if ( (x - 1) < 0 && (y + 1) < mat[0].length) {
                    y++; // go right
                }
        
                // case: if you can't do any of the above, go down 
                else {
                    x++; // go down
                }
            }
            // case2: if x + y is odd
            else {
        
                // case: if you can go down and left
                if ( (x + 1) < mat.length && (y - 1) >= 0 ) {
                    x++; // go down
                    y--; // go left
                }
        
                // case: if you can go down but not left
                else if ( (x + 1) < mat.length && (y - 1) < 0 ) {
                    x++; // go down
                }
        
                // case: if can't do any of the above, go right
                else {
                    y++; // go right
                }
            }
        }
        // return result
        return result;
        
    }
}